### PR TITLE
Feat(framework): Create L2 contract Move interfaces

### LINKED ISSUE #67: Review Optimism L2 Solidity contracts
We need to know if any of these contracts need to be translated to Move and included in our genesis process. They will be needed if they provide functionality which we also need.

### PR DESCRIPTION
**Pull request by bulenttastan**
### Description
Automatically generate L2 contract interfaces on the Move side to interact with the pre-deployed L2 Solidity contracts. Closes #67 

### Changes
The main changes are in [generate](https://github.com/MovedNetwork/op-move/pull/175/files#diff-deab4a6494fc2a4635cb94ac3fe102522a550c33bee4ac0861f3853752d06e1a) and [solidity_abi](https://github.com/MovedNetwork/op-move/pull/175/files#diff-8b50ca4361134c97969946eb1c845f373549a9f258558dc955d073c2fdf71b1e) files.
- Collect the relevant data to pass into file writer
- Write to actual Move files using a Move module template
- Filter out L1 contracts and convert only L2 contracts
- Remove the redundant `L2CrossDomainMessenger` module
- Test out `abi_decode_params` native extension

**Missing Work**
- More comprehensive testing of L2 scripts
- (Optional) Execute EVM call without a signer for no state change methods

### Testing
✓ Generated file matches the existing `L2CrossDomainMessenger` module
✓ Simple test to read the `version` method on an L2 contract

### CODE REVIEW

**Review by birchmd — APPROVED** [review]
Awesome work on this!

File: genesis-builder/l2_move_template.hbs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    struct {{pascal name}}Args {}
+    {{/if}}
+
+    public fun {{snake name}}(
+        caller: &signer,
+        {{#each inputs}}
+        {{snake name}}: {{{ty}}},
+        {{/each}}
+        {{#if has_value}}
+        _value: FungibleAsset,
```
[the remark this answers is not in the document]
Comment (bulenttastan) [reply genesis-builder/l2_move_template.hbs:35]: Some contract methods have parameters named `value`, not necessarily meaning account value. Some are even named `_value`. The code generation always removes the initial underscores from input params to have a standard naming convention.
So it's safe to make internal params start with underscores.

File: moved/src/move_execution/evm_native/solidity_abi.rs [hunk 2]
```diff
@@ -175,6 +221,95 @@ fn move_value_to_sol_value(mv: MoveValue, annotated_layout: &MoveTypeLayout) ->
...
+        }
+        MoveTypeLayout::Struct(struct_layout) => {
+            let (struct_tag, field_layouts) = match struct_layout {
+                MoveStructLayout::WithTypes { type_, fields } => (type_, fields),
+                _ => unreachable!("Must have type because layout is constructed with `type_to_fully_annotated_layout`"),
+            };
+
+            // Special case data marked as being fixed bytes
+            if FIXED_BYTES_TAG.eq(struct_tag) {
+                // TODO: Support any `bytesN` fixed bytes. Possibly need a struct for each one in Evm
```
[the remark this answers is not in the document]
Comment (bulenttastan) [reply moved/src/move_execution/evm_native/solidity_abi.rs:249]: That's also a good idea, like `Byte4`, `Byte32` enums. I'll add that idea to the comment too to help with the future PR.

File: moved/src/move_execution/evm_native/solidity_abi.rs [hunk 3]
```diff
@@ -175,6 +221,95 @@ fn move_value_to_sol_value(mv: MoveValue, annotated_layout: &MoveTypeLayout) ->
...
+
+            // All other struct types are tuples in Solidity
+            DynSolType::Tuple(
+                field_layouts
+                    .iter()
+                    .map(|field_layout| layout_to_sol_type(&field_layout.layout))
+                    .collect(),
+            )
+        }
+        MoveTypeLayout::Native(_, _) => unreachable!("Native move layout is not supported"),
```
[the remark this answers is not in the document]
Comment (bulenttastan) [reply moved/src/move_execution/evm_native/solidity_abi.rs:277]: Will look more into `MoveTypeLayout::Native` to fully understand whether it's ever used or not.

File: moved/src/move_execution/evm_native/solidity_abi.rs [hunk 4]
```diff
@@ -175,6 +221,95 @@ fn move_value_to_sol_value(mv: MoveValue, annotated_layout: &MoveTypeLayout) ->
...
+            // Instead of equality, check specifically for module id and name to match.
+            if FIXED_ARRAY_TAG.module_id() == struct_tag.module_id()
+                && FIXED_ARRAY_TAG.name == struct_tag.name
+            {
+                // Fixed size vs dynamic array are defined e.g. as `uint[3]` vs `uint[]` in Solidity.
+                // Move doesn't support fixed size vectors, hence we don't know the actual size beforehand.
+                unimplemented!("Fixed size arrays are not supported in move");
+            }
+
+            if STRING_TAG.eq(struct_tag) {
```
[the remark this answers is not in the document]
Comment (bulenttastan) [reply moved/src/move_execution/evm_native/solidity_abi.rs:265]: Apparently `LazyLock`ed variables cannot be compared directly. Direct comparison gives an error:
```
Binary operation == cannot be applied to type LazyLock<StructTag, fn() -> StructTag> [E0369]
```

File: moved/src/move_execution/evm_native/solidity_abi.rs [hunk 5]
```diff
@@ -175,6 +221,95 @@ fn move_value_to_sol_value(mv: MoveValue, annotated_layout: &MoveTypeLayout) ->
...
+            DynSolType::Array(Box::new(layout_to_sol_type(vector_layout)))
+        }
+        MoveTypeLayout::Struct(struct_layout) => {
+            let (struct_tag, field_layouts) = match struct_layout {
+                MoveStructLayout::WithTypes { type_, fields } => (type_, fields),
+                _ => unreachable!("Must have type because layout is constructed with `type_to_fully_annotated_layout`"),
+            };
+
+            // Special case data marked as being fixed bytes
+            if FIXED_BYTES_TAG.eq(struct_tag) {
```
[the remark this answers is not in the document]
Comment (bulenttastan) [reply moved/src/move_execution/evm_native/solidity_abi.rs:248]: Unfortunately it gives `LazyLock` comparison error for direct `==` check.

**Pull request merged**